### PR TITLE
Introduce smart price search for Catalog Price rule query builder

### LINKED ISSUE #15890: How to handle search in reduction column for "Catalog > Discounts > Catalog price rules > Add new / edit catalog price rule" page
We have migrated page "Catalog > Discounts > Catalog price rules > Add new / edit catalog price rule" in PR https://github.com/PrestaShop/PrestaShop/pull/13584

However we have found one issue: https://github.com/PrestaShop/PrestaShop/pull/13584#discussion_r284236594

Basically, what happens is:
1) I create a catalog price rule with a reduction of 10.00005
2) I go to listing: well, on the listing the reduction is displayed but *rounded* so I see it was 10.00

Capture d’écran 2019-10-09 à 18 27 36

3) However ... the "reduction" input I can use to "search for catalog price rules" is looking for an exact match.
=> So if I search all catalog rules whose reduction is 10.00005, I find my catalog price rule
=> So if I search all catalog rules whose reduction is 10.0, I **cannot find** my catalog price rule

There's a weird behavior here, which is that:
- the "search" input looks for exact matches
- but the data is being displayed after a rounding, so the customer cannot see the exact match 😄 

@MatShir @marionf

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Allow Catalog Price rule query builder to consider number of decimals used to search for prices and truncate accordingly search dataset, allowing "smart" searches. See scenario below to fully understand.
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/15890
| How to test?  | See below

## How to test

Go on migrated (hidden) page `admin-dev/index.php/sell/catalog/catalog-price-rules`.
Create 2 catalog price rules: one with reduction of 10.0 and the other one with reduction of 10.1.
Before this PR, in listing, try to search using "reduction" field with value 10. You will only see 1 result: the row with 10.0 . This is because the code considers 10 is equal to 10.000000 .
With this PR
- if you search "10" you will see results 10.0 and 10.1
- if you search "10.1" you will see result 10.1
- if you search "10.0" you will see result 10.0

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/18691)

### CODE REVIEW

**Review by Progi1984 — DISMISSED** [review]

File: src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php [hunk 1]
```diff
@@ -203,8 +200,41 @@ private function applyFilters(QueryBuilder $qb, array $filters)
...
+    private function findNumberOfDecimals($value): int
+    {
+        if (!is_string($value)) {
+            throw new \InvalidArgumentException('Expected string');
+        }
+
+        $numberOfDecimals = 0;
+        $explodedValue = explode('.', $value);
+
+        if (count($explodedValue) > 1) {
```
Comment (PierreRambaud) [inline src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php]:
```suggestion
        if (isset($explodedValue[1])) {
```
Don't count the whole array if you only need the `[1]` index

File: src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php [hunk 2]
```diff
@@ -203,8 +200,41 @@ private function applyFilters(QueryBuilder $qb, array $filters)
...
+
+    /**
+     * @param string $value
+     *
+     * @return int
+     */
+    private function findNumberOfDecimals($value): int
+    {
+        if (!is_string($value)) {
+            throw new \InvalidArgumentException('Expected string');
```
Comment (PierreRambaud) [inline src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php]:
```suggestion
            throw new InvalidArgumentException('Expected string');
```
use namespace

**Review by Progi1984 — COMMENTED** [review]
I approve @PierreRambaud Feedback

**Comment by matks** [comment]
Code review applied in https://github.com/PrestaShop/PrestaShop/commit/208dba01ae89bd3d39e139a230a67e500f83b5f1

**Review by Progi1984 — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Comment by matks** [comment]
@PrestaShop/qa-functional all right, you guys have to tell me ... did I say something harsh ? 😢 you don't want to test my PR

**Comment by Progi1984** [comment]
@matks Please wait, the @PrestaShop/qa-functional team has a lot of [waiting PR](https://github.com/pulls?q=is%3Aopen+is%3Apr+archived%3Afalse+label%3A%22Waiting+for+QA%22+org%3APrestaShop)

**Comment by matks** [comment]
> @matks Please wait, the @PrestaShop/qa-functional team has a lot of [waiting PR](https://github.com/pulls?q=is%3Aopen+is%3Apr+archived%3Afalse+label%3A%22Waiting+for+QA%22+org%3APrestaShop)

You're a nice person when you defend them but I know the truth 😢 @Robin-Fischer-PS is disappointed I dont drink beer with him anymore because I'm not in the office

**Comment by Progi1984** [comment]
:beers: <3

**Comment by Robin-Fischer-PS** [comment]
@matks 

Never forget Suktinis...

IMG_20200114_181615

**Comment by SD1982** [comment]
LGTM thanks @matks !

**Comment by matks** [comment]
Thank you for QA and review !

**Pull request merged**